health/health.go

```diff
@@ -1159,14 +1180,10 @@ func (t *Tracker) updateBuiltinWarnablesLocked() {
 		t.setHealthyLocked(derpRegionErrorWarnable)
 	}
 
-	if len(t.controlHealth) > 0 {
-		for _, s := range t.controlHealth {
-			t.setUnhealthyLocked(controlHealthWarnable, Args{
-				ArgError: s,
-			})
-		}
-	} else {
-		t.setHealthyLocked(controlHealthWarnable)
+	// Check if control health messages have changed
+	if !slices.EqualFunc(t.lastControlMessages, t.controlMessages, tailcfg.DisplayMessage.Equal) {
```

Comment:
If watchers really care about whether there is a diff in control messages, they're still going to have to check themselves. Though we're minimising the number of necessary invocations of the watchers, they're spun up in their own goroutine and you could end up with a sequence like:

```
t.SetControlHealth([]tailcfg.DisplayMessage{{ID: "a"}})
go watcher(nil, nil)
t.SetControlHealth([]tailcfg.DisplayMessage{{ID: "b"}})
go watcher(nil, nil)

CurrentState() => "b"
CurrentState() => "b"
```

Reply:
Why do we have the diff at all, then?

Reply:
One watcher works on diffs, and one doesn't. The watcher that reports Warnables to the control plane cares about diffs, and that one gets the change as an argument to it, so it will get each diff and if a Warnable flaps like your example above it will report it going both on and off (the arguments would not be nil in the case of a Warnable, not a ControlHealth message, changing).

The watcher that needs to know something has changed to send the overall health state to the UI doesn't need the diff, we just want the UI to display the latest current state. So it's fine for that one to just fetch the end state and if something very briefly flaps it won't show in the UI.